fix(lint): address unchecked errors

antoineco opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

  Kraftkit follows the same guidelines as the Unikraft Open Source Project.

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [x] Updated relevant documentation.

### Description of changes

Part of #413
Closes #421

I'll be gradually enabling golangci-lint's default linters, starting with the one reporting the most issues in kraftkit: `errcheck`.

antoineco committed:
fix(lint): address unchecked errors

Signed-off-by: Antoine Cotten <antoine@unikraft.io>

antoineco commented:
@nderjung It looks like these changes uncovered a previously hidden issue:
```
panic: invalid argument "<bool Value>" for "--fast" flag: strconv.ParseBool: parsing "<bool Value>": invalid syntax

goroutine 1 [running]:
kraftkit.sh/cmd/kraft/build.New()
	/home/acotten/git/unikraft/kraftkit/cmd/kraft/build/build.go:71 +0x20d
main.New()
	/home/acotten/git/unikraft/kraftkit/cmd/kraft/kraft.go:69 +0x28a
main.main()
	/home/acotten/git/unikraft/kraftkit/cmd/kraft/kraft.go:101 +0x2b
```

nderjung commented:
> @nderjung It looks like these changes uncovered a previously hidden issue:
> 
> ```
> panic: invalid argument "<bool Value>" for "--fast" flag: strconv.ParseBool: parsing "<bool Value>": invalid syntax
> 
> goroutine 1 [running]:
> kraftkit.sh/cmd/kraft/build.New()
> 	/home/acotten/git/unikraft/kraftkit/cmd/kraft/build/build.go:71 +0x20d
> main.New()
> 	/home/acotten/git/unikraft/kraftkit/cmd/kraft/kraft.go:69 +0x28a
> main.main()
> 	/home/acotten/git/unikraft/kraftkit/cmd/kraft/kraft.go:101 +0x2b
> ```

@craciunoiuc you recently looked into this issue (i.e. correctly handling `--fast`), can you take a look?

craciunoiuc commented:
Yup sure

Edit: I did not touch this at this stage, but are you maybe calling it like `--fast true`/`--fast false`? It should be `--fast`/` `

antoineco commented:
@craciunoiuc we are not executing the command with any flag.

The error occurs during the initialization of the `Cmd`:

https://github.com/unikraft/kraftkit/blob/fbe71fd12aa075d0b1751ca13bedd118de202030/cmd/kraft/version/version.go#L20-L31

craciunoiuc commented:
I can take a deeper look once #408 is in because it changed my environment a bit and it's hard to switch back and forth

Right now I have no idea why it's not working

antoineco commented:
@nderjung this is actually a global issue with boolean values.
It's also not too clear to me how a flag defined in `cmd/kraft/build` made its way in `cmd/kraft/version`.

It's failing here:

https://github.com/unikraft/kraftkit/blob/fbe71fd12aa075d0b1751ca13bedd118de202030/cmdfactory/builder.go#L198-L202

- The value of `name` is `"fast"`
- The value of `strValue` is `"<bool Value>"` (:exclamation:)

Inside `flags.Set`, `flag.Value.Set(strValue)` eventually calls `strconv.ParseBool(strValue)`, which fails with:

```
error(*strconv.NumError) *{
        Func: "ParseBool",
        Num: "<bool Value>",
        Err: error(*errors.errorString) *{
                s: "invalid syntax",},}
```

craciunoiuc commented:
Well, all commands get initialized in the beginning, including build, `fast` is just the first one to trigger the fail

The error happens on all commands

craciunoiuc commented on oci/handler/containerd.go:
```diff
@@ -103,21 +105,21 @@ func (handle *ContainerdHandler) DigestExists(ctx context.Context, dgst digest.D
 	}
 
-	defer done(ctx)
+	defer func() {
+		err = closeLease(ctx, done, err)
+	}()
```
Does this alter the behaviour?

antoineco replied:
It shouldn't.
The helper executes the `done` function like it used to, but now also does some error mangling to make sure that all errors are handled (and combined if necessary).

antoineco replied:
I could have also written it like this:

```go
	defer func() {
		err = combineErrors(done(ctx), err)
	}()
```

Maybe that would be more explicit, actually.

craciunoiuc replied:
ah, okay, makes sense, it got me confused about leases. We can leave it like this then

craciunoiuc replied:
ah you already changed, nice :D

antoineco replied:
Done in [709f3ea9..c8d3a91f](https://github.com/unikraft/kraftkit/compare/709f3ea93c82a8922069d3cec19507c1e7dd13af..c8d3a91f09828b9c40a6154e77e5e108cc000f6c)

craciunoiuc reviewed (approved):
All good here. Thanks!

Looks fine from my side, but I'll let Alex have the final say.

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

antoineco commented:
Something broke. `kraft version --help` now exits with 1 and print:

```console
$ kraft version --help
pflag: help requested
```

Investigating.

antoineco committed:
fix(cmd): ignore pflag.ErrHelp while parsing flags

Avoids exiting prematurely when a command's help is requested.

GitHub-Fixes #421

Signed-off-by: Antoine Cotten <antoine@unikraft.io>

antoineco commented:
I found the issue. See my last commit.
Tests should pass now.
